### ISSUE TITLE
Time-travel debugging - follow-up

### ISSUE THREAD

**Issue by signor-pedro** [patchlevel/event-sourcing#633]
Hi guys, once again, thanks for this fantastic library. 👍 

I would like to follow up on our previous discussion #619 where we already learned how to easily restore an aggregate to any previous point in history.

However, what if some of my data aren't stored on the aggregate, only projected from the events?

Imagine we have an entity `User`. The entity has some "displayable" properties such as `name`, `avatar` etc. that are rendered in the UI but in no way affect the state machine and its transitions. 

I tend not to store them on the aggregate, only project them, because I read somewhere that you should only store data **relevant to state transition** on the aggregate, e.g.:
```php
// UserAggregate.php

public function ban(BanUser $event): void
    {
        if ($this->isBanned) {
            throw new \LogicException("User is already banned.");
        }

        $this->recordThat($event);
    }
```

In other words, with such "slim aggregates" I can time-travel to whether the user was banned 3 days ago, but I would not read what the users avatar was 3 days ago from the aggregate.

How do you deal with this?
- is it advised to store **everything** on the aggregate, kind of against the ES "best practices"?
- or is it advised to build some mechanism like "DebuggingProjector" that would somehow run projections on the limited subset of events, and instead of actually perform database updates, would run those "in-memory" and dump the output entity to the console etc?

I guess both is doable but I would like to hear from ES masters what their professional advice is 🤓 

Thanks!

**Comment by DavidBadura** [comment]
Hey,

yes, you can of course also build a temporary in-memory projection. In our use case (admin UI), the goal was to debug the state to understand the values present in the aggregate so that you could understand why certain events were thrown.

But in your use case it makes sense to simply build a different object instead. The principle remains the same here:

1) Get a stream with the necessary events
2) Consume events and build a state

It doesn't matter whether this is an aggregate or a projection.

I've added an example here to use a projection for this, but you could theoretically build something yourself.

```php
use Patchlevel\EventSourcing\Aggregate\AggregateHeader;
use Patchlevel\EventSourcing\Attribute\Subscribe;
use Patchlevel\EventSourcing\EventBus\DefaultConsumer;
use Patchlevel\EventSourcing\Message\Message;
use Patchlevel\EventSourcing\Store\Criteria\AggregateIdCriterion;
use Patchlevel\EventSourcing\Store\Criteria\AggregateNameCriterion;
use Patchlevel\EventSourcing\Store\Criteria\Criteria;

$userViewer = new class()
{
    public string $name;

    #[Subscribe(UserCreated::class)]
    public function userCreated(Message $message): void
    {
        $this->name = $message->event()->name;
    }
};

$consumer = DefaultConsumer::create([$userViewer]);

$criteria = new Criteria(
    new AggregateNameCriterion($aggregateName),
    new AggregateIdCriterion($aggregateId),
);

$stream = $this->store->load($criteria);

$until = new \DateTimeImmutable('2021-01-01');

foreach ($stream as $message) {
    if ($message->header(AggregateHeader::class)->recordedOn > $until) {
        break;
    }

    $consumer->consume($message);
}

echo $userViewer->name;
```

I notice that this use case of creating a temporary view is quite common.
Maybe we could build a nicer API for this.

**Comment by DavidBadura** [comment]
Maybe an API like this:

```php
use Patchlevel\EventSourcing\Message\Translator\UntilEventTranslator;
use Patchlevel\EventSourcing\Store\Criteria\AggregateIdCriterion;
use Patchlevel\EventSourcing\Store\Criteria\AggregateNameCriterion;

$result = (new Projection($this->store))
    ->init([
        'name' => 'Unknown',
    ])
    ->criteria(
        new AggregateNameCriterion('user'),
        new AggregateIdCriterion('123'),
    )
    ->translator(new UntilEventTranslator(new DateTimeImmutable('2021-01-01')))
    ->when(UserCreated::class, fn(array $state, UserCreated $event) => [
        'name' => $event->name,
        ...$state,
    ])
    ->run();

echo $result['name'];
```

**Comment by signor-pedro** [comment]
Thank you for the detailed explanation. 

Makes sense - the API you propose would be convenient, but we can work with a custom viewer. You are right that in the end it is easy to get the state - just loop through the events and build it up again.

Perhaps I complicated it too much thinking we need to approach this from the infrastructure perspective (replace `EntityManager` with some in-memory object tracker injected in a debugging run of the projection, so that you do not have to write a custom viewer). But I gave it extra thought and that would open far too many edgecases for something as simple as building variable's state.

Thanks.
